dvc/repo/ls.py

```diff
@@ -0,0 +1,90 @@
+import os
+
+from dvc.exceptions import DvcException, OutputNotFoundError
+
+
+@staticmethod
+def ls(url, target=None, rev=None, recursive=None, outputs_only=False):
+    from dvc.external_repo import external_repo
+    from dvc.repo import Repo
+
+    with external_repo(url, rev) as repo:
+        result = set()
+        if isinstance(repo, Repo):
+            result.update(_ls_outs_repo(repo, target, recursive))
+        if not outputs_only:
+            try:
+                result.update(_ls_files_repo(repo, target, recursive))
+            except DvcException as exc:
+                if not len(result):
+                    raise exc
```

Comment:
Empty output from `_ls_outs_repo` doesn't mean that this is an error, as it swallows OutputNotFoundError so in your code you are not distinguishing between "no output found" and "output directory is empty" cases. Please take a look at `dvc get`, or more specifically at `ExternalRepo.pull_to`, it has a lot of similar logic that you can apply here.

Reply:
Oh, though I didn't notice that if the dir is empty we are still returning ["empty_dir"], so this does work. :+1: